src/festim/coupled_heat_hydrogen_problem.py

```diff
@@ -0,0 +1,200 @@
+    another function (f_out) with a mismatching mesh
+
+    args:
+        f_out: function to interpolate into
+        f_in: function to interpolate from
+
+    """
+
+    dim = f_out.function_space.mesh.topology.dim
+    index_map = f_out.function_space.mesh.topology.index_map(dim)
+    ncells = index_map.size_local + index_map.num_ghosts
+    cells = np.arange(ncells, dtype=np.int32)
+    interpolation_data = fem.create_interpolation_data(
+        f_out.function_space, f_in.function_space, cells, padding=1e-11
+    )
+    f_out.interpolate_nonmatching(f_in, cells, interpolation_data=interpolation_data)
+
+
+class CoupledHeatTransferHydrogenTransport:
+    """
+    Coupled heat transfer and hydrogen transport problem
+
+    Args:
+        heat_problem: the heat transfer problem
+        hydrogen_problem: the hydrogen transport problem
+
+    Attributes:
+        heat_problem: the heat transfer problem
+        hydrogen_problem: the hydrogen transport problem
+        non_matching_meshes: True if the meshes in the heat_problem and hydorgen_problem
+            are not matching
+
+    Examples:
+        .. highlight:: python
+        .. code-block:: python
+
+            import festim as F
+
+            my_heat_transfer_model = F.HeatTransferProblem(
+                mesh=F.Mesh(...),
+                subdomains=[F.Subdomain(...)],
+                ...
+            )
+
+            my_h_transport_model = F.HydrogenTransportProblem(
+                mesh=F.Mesh(...),
+                subdomains=[F.Subdomain(...)],
+                species=[F.Species(name="H"), F.Species(name="Trap")],
+                ...
+            )
+
+            coupled_problem = F.CoupledHeatTransferHydrogenTransport(
+                heat_problem=my_heat_transfer_model,
+                hydrogen_problem=my_h_transport_model,
+            )
+
+
+    """
+
+    heat_problem: HeatTransferProblem
+    hydrogen_problem: HydrogenTransportProblem
+
+    non_matching_meshes: bool
+
+    def __init__(
+        self,
+        heat_problem: HeatTransferProblem,
+        hydrogen_problem: HydrogenTransportProblem,
+    ) -> None:
+        self.heat_problem = heat_problem
+        self.hydrogen_problem = hydrogen_problem
+
+    @property
+    def heat_problem(self):
+        return self._heat_problem
+
+    @heat_problem.setter
+    def heat_problem(self, value):
+        if not isinstance(value, HeatTransferProblem):
+            raise TypeError("heat_problem must be a festim.HeatTransferProblem object")
+        self._heat_problem = value
+
+    @property
+    def hydrogen_problem(self):
+        return self._hydrogen_problem
+
+    @hydrogen_problem.setter
+    def hydrogen_problem(self, value):
+        if not isinstance(value, HydrogenTransportProblem):
+            raise TypeError(
+                "hydrogen_problem must be a festim.HydrogenTransportProblem object"
+            )
+        self._hydrogen_problem = value
```

Comment:
I believe this coupling won't work with the change of variable method for multi-materials since it needs the temperature at the previous time step. Should maybe just check that `value` is not of type `HydrogenTransportProblemDiscontinuousChangeVar`

Reply:
ok will do